File: julia/Euler/src/bin/p0060.jl

# project euler: problem 60

#=

size of clique | primes
---------------+--------------------------------------
       4       | 3, 7, 109, 673
       5       | 13, 5197, 5701, 6733, 8389
       6       | 25819, 29569, 209623, 234781, 422089, 452041


Below are results on Raspberry Pi 4 and Mac mini 2018 (i5-8500B/3GHz).

# Raspberry Pi 4

julia> versioninfo()
Julia Version 1.10.2
Commit bd47eca2c8a (2024-03-01 10:14 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (aarch64-linux-gnu)
  CPU: 4 × Cortex-A72
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, cortex-a72)
Threads: 1 default, 0 interactive, 1 GC (on 4 virtual cores)

julia> @time solve_0060(5)
  1.770793 seconds (159.93 k allocations: 12.773 MiB)
26033

julia> @time solve_0060(4)
  0.004136 seconds (2.12 k allocations: 151.422 KiB)
792


# Mac mini 2018

julia> versioninfo()
Julia Version 1.10.2
Commit bd47eca2c8a (2024-03-01 10:14 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: macOS (x86_64-apple-darwin22.4.0)
  CPU: 6 × Intel(R) Core(TM) i5-8500B CPU @ 3.00GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, skylake)
Threads: 1 default, 0 interactive, 1 GC (on 6 virtual cores)

julia> @time solve_0060(6)
3443.381419 seconds (565.45 M allocations: 38.210 GiB, 0.12% gc time)
1373922

julia> @time solve_0060(5)
  0.580957 seconds (159.93 k allocations: 12.773 MiB)
26033

julia> @time solve_0060(4)
  0.001300 seconds (2.12 k allocations: 151.422 KiB)
792

=#

module Prob0060

import Primes: isprime, nextprime

function get_pairable_primes(x, asc_prime_lst, limit)
    function is_pair(x, upper_x, y, upper_y)
        isprime(x * upper_y + y) && isprime(y * upper_x + x)
    end
    upper_x = 10 ^ ndigits(x)
    upper_p = 10
    result = Int[]
    for p in asc_prime_lst
        if x + p >= limit
            break
        end
        while p > upper_p
            upper_p *= 10
        end
        if is_pair(x, upper_x, p, upper_p) == true
            push!(result, p)
        end
    end
    result
end

function find_cliques(desc_prime_lst, size, tbl)
    function aux(group, offset, depth)
        if depth == 0
            push!(result, group)
        else
            for idx in offset:(length(desc_prime_lst) - depth + 1)
                if all(x -> in(desc_prime_lst[idx], tbl[x]), group)
                    aux(push!(copy(group), desc_prime_lst[idx]), idx + 1, depth - 1)
                end
            end
        end
    end

    result = Vector{Int}[]
    aux(Array{Int,1}(undef,0), 1, size)
    result
end

function solve_0060(size_of_clique::Int = 5)
    @assert size_of_clique > 1 "invalid parameter"
    prime_lst = [[3], [3]]    # Group prime numbers by the remainder divided by 3 (but include 3).
    tbl = Dict{Int, Set{Int}}()
    answer = typemax(Int)

    # start from the 4th prime, 7
    p = 5
    while true
        p = nextprime(p + 1)

        # break this loop when it has verified the answer is smallest
        if p >= answer
            break
        end

        # find all prime numbers smaller than 'p' that can be paired with 'p'
        idx = p % 3
        nbr_lst = get_pairable_primes(p, prime_lst[idx], answer)
        tbl[p] = Set(nbr_lst)
        # update known prime numbers
        push!(prime_lst[idx], p)

        # if number of connectable primes is less than 'size_of_clique - 1', check the next prime.
        if length(nbr_lst) < size_of_clique - 1
            continue
        end

        cliques = find_cliques(reverse(nbr_lst), size_of_clique - 1, tbl)
        if isempty(cliques) == false
            answer = min(answer, minimum(map(x -> p + sum(x), cliques)))
        end
    end
    answer
end

end #module

using .Prob0060: solve_0060
export solve_0060

File: rust/src/bin/p0060.rs
// Project Euler: Problem 60

use euler::math::primes;
use std::collections::{HashMap, HashSet};

euler::run_solver!(60);

fn solve() -> String {
    compute(5).to_string()
}

fn compute(size_of_clique: usize) -> u64 {
    use std::cmp;

    debug_assert!(size_of_clique > 1);

    let mut prime_lst: Vec<Vec<u64>> = vec![vec![3], vec![3]];
    let mut tbl: HashMap<u64, HashSet<u64>> = HashMap::new();
    let mut ans = u64::MAX;

    // start searching from the 4th prime, 7
    let mut p: u64 = 5;
    loop {
        // break this loop when it has verified the answer is the lowest sum
        p = primes::next_prime(p);
        if p >= ans {
            break;
        }

        // search for prime numbers that can be paired with 'p' and are less than 'p'
        let idx = ((p % 3) - 1) as usize;
        let mut nbr_lst = find_nbrs(p, &prime_lst[idx], ans);
        tbl.insert(p, HashSet::from_iter(nbr_lst.clone()));

        // update checked prime numbers
        prime_lst[idx].push(p);

        // if number of connectable primes is less than 'size_of_clique - 1', check the next larger prime.
        if nbr_lst.len() < size_of_clique - 1 {
            continue;
        }

        // switch the list of connectable primes to descending order and check cliques.
        nbr_lst.reverse();
        if let Some(clqs) = check_cliques(&nbr_lst, size_of_clique - 1, &tbl) {
            ans = cmp::min(
                ans,
                clqs.into_iter()
                    .map(|v| v.into_iter().sum::<u64>() + p)
                    .min()
                    .unwrap(),
            );
        }
    }

    ans
}

fn is_pair(x: u64, y: u64) -> bool {
    fn concat(a: u64, b: u64) -> u64 {
        let mut n: u64 = 10;
        while b > n {
            n *= 10;
        }
        a * n + b
    }

    primes::is_prime(concat(x, y)) && primes::is_prime(concat(y, x))
}

fn find_nbrs(p: u64, asc_p_lst: &[u64], current_ans: u64) -> Vec<u64> {
    asc_p_lst
        .iter()
        .take_while(|x| **x < current_ans - p)
        .filter(|x| is_pair(**x, p))
        .copied()
        .collect()
}

fn check_cliques(
    desc_nbr_lst: &[u64],
    size: usize,
    tbl: &HashMap<u64, HashSet<u64>>,
) -> Option<Vec<Vec<u64>>> {
    fn aux(
        group: &[u64],
        ps: &[u64],
        depth: usize,
        tbl: &HashMap<u64, HashSet<u64>>,
        result: &mut Vec<Vec<u64>>,
    ) {
        if depth == 0 {
            result.push(group.to_vec());
        } else {
            for offset in 0..=(ps.len() - depth) {
                if group
                    .iter()
                    .all(|x| tbl.get(x).unwrap().contains(&ps[offset]))
                {
                    let mut next_group = group.to_vec();
                    next_group.push(ps[offset]);
                    aux(&next_group, &ps[(offset + 1)..], depth - 1, tbl, result);
                }
            }
        }
    }

    let mut result: Vec<Vec<u64>> = Vec::new();
    aux(&Vec::new(), desc_nbr_lst, size, tbl, &mut result);

    if result.is_empty() {
        None
    } else {
        Some(result)
    }
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0060_3() {
        // 3, 37, 67
        assert_eq!(compute(3), 107);
    }

    #[test]
    fn p0060_4() {
        // 3, 7, 109, 673
        assert_eq!(compute(4), 792);
    }

    #[test]
    fn p0060_5() {
        // 13, 5197, 5701, 6733, 8389
        assert_eq!(compute(5), 26033);
    }
}
